clients/src/main/java/org/apache/kafka/clients/Metadata.java

```diff
@@ -50,6 +50,7 @@
 
     private final long refreshBackoffMs;
     private final long metadataExpireMs;
+    private final long metadataStaleMs;
```

Comment:
Since the metadataStaleMs will be checked for each batch, it probably makes sense to have a separate variable. The other comments are still my concerns.